--- /tmp/rfcommands/rfcommands/cli/compile_step_stats.py	2025-09-22 13:35:59.826336638 -0500
+++ modified_rfc_tools/cli_compile_step_stats.py	2025-09-22 13:33:38.024654155 -0500
@@ -35,8 +35,13 @@
               help     = "Name of the experiment." ,
               required = True,
               type     = click.STRING)
-def compile_step_stats(out,     cutadapt, filter, trans, 
-                       quality,   dedup,  name):
+@click.option('--label-prefix', '-l',
+              help     = "Label prefix for alignment statistics (transcriptome or genome)." ,
+              required = False,
+              default  = "transcriptome",
+              type     = click.Choice(['transcriptome', 'genome']))
+def compile_step_stats(out,     cutadapt, filter, trans,
+                       quality,   dedup,  name, label_prefix):
     """
     Puts statistics coming from various steps into one file.
     
@@ -53,10 +58,11 @@
     you can re-calculate using the counts given in these tables.
     """
 
-    compile(out      = out,
-            cutadapt = cutadapt, 
-            filter   = filter, 
-            trans    = trans, 
-            quality  = quality,  
-            dedup    = dedup,  
-            name     = name)
+    compile(out          = out,
+            cutadapt     = cutadapt,
+            filter       = filter,
+            trans        = trans,
+            quality      = quality,
+            dedup        = dedup,
+            name         = name,
+            label_prefix = label_prefix)
